annotated.go

```diff
@@ -89,3 +90,228 @@ func (a Annotated) String() string {
+// for annotating the parameter and result types of a function.
+type Annotation interface {
+	getAnnotatedType(reflect.Type) []reflect.Type
+}
+
+type paramTags struct {
+	tags []string
+}
+
+func (p paramTags) getAnnotatedType(fType reflect.Type) []reflect.Type {
+	annotatedParams := []reflect.StructField{{
+		Name:      "In",
+		Type:      reflect.TypeOf(In{}),
+		Anonymous: true,
+	}}
+
+	for i := 0; i < fType.NumIn(); i++ {
+		structField := reflect.StructField{
+			Name: fmt.Sprintf("Field%d", i),
+			Type: fType.In(i),
+		}
+		if i < len(p.tags) {
+			structField.Tag = reflect.StructTag(p.tags[i])
+		}
+		annotatedParams = append(annotatedParams, structField)
+	}
+
+	paramTypes := []reflect.Type{reflect.StructOf(annotatedParams)}
+	return paramTypes
+}
+
+// ParamTags is an Annotation that annotates the parameter(s) of a function.
+// When multiple tags are specified, each tag is mapped to the corresponding
+// positional parameter.
+func ParamTags(tags ...string) Annotation {
+	p := paramTags{tags}
+	return p
+}
+
+type resultTags struct {
+	tags []string
+}
+
+func (resultTags) getAnnotatedType(fType reflect.Type) []reflect.Type {
+	annotatedResult := []reflect.StructField{{
+		Name:      "Out",
+		Type:      reflect.TypeOf(Out{}),
+		Anonymous: true,
+	}}
+
+	for i := 0; i < fType.NumOut(); i++ {
+		structField := reflect.StructField{
+			Name: fmt.Sprintf("Field%d", i),
+			Type: fType.Out(i),
+		}
+		annotatedResult = append(annotatedResult, structField)
+	}
+
+	resultTypes := []reflect.Type{reflect.StructOf(annotatedResult)}
+	return resultTypes
+}
+
+// ResultTags is an Annotation that annotates the result(s) of a function.
+// When multiple tags are specified, each tag is mapped to the corresponding
+// positional result.
+func ResultTags(tags ...string) Annotation {
+	r := resultTags{tags}
+	return r
+}
+
+// Annotate lets you annotate a function's paramter and returns with tags
+// without you having to declare separate struct definitions for them.
+//
+// For example,
+//   func NewGateway(ro, rw *db.Conn) *Gateway { ... }
+//   fx.Provide(
+//     fx.Annotate(
+//       NewGateway,
+//       fx.ParamTags(`name:"ro" optional:"true"`, `name:"rw"`),
+//       fx.ResultTags(`name:"foo"`),
+//     ),
+//   )
+//
+// Is equivalent to,
+//
+//  type params struct {
+//    fx.In
+//
+//    RO *db.Conn `name:"ro" optional:"true"`
+//    RW *db.Conn `name:"rw"`
+//  }
+//
+//  type result struct {
+//    fx.Out
+//
+//    GW *Gateway `name:"foo"`
+//   }
+//
+//   fx.Provide(func(p params) result {
+//     return result{GW: NewGateway(p.RO, p.RW)}
+//   })
+//
+// If there are more Annotations are specified than than one per
+// Annotation type, none of the Annotations will be applied.
+//
+// For example,
+//
+// fx.Provide(
+//   fx.Annotate(
+//     NewGateWay,
+//     fx.ParamTags(`name:"ro" optional:"true"`),
+//     fx.ParamTags(`name:"rw"),
+//     fx.ResultTags(`name:"foo"`)
+//   )
+// )
```

Comment:
Code samples should be indented at least one more space for godoc to render them as code blocks.
```suggestion
//  fx.Provide(
//    fx.Annotate(
//      NewGateWay,
//      fx.ParamTags(`name:"ro" optional:"true"`),
//      fx.ParamTags(`name:"rw"),
//      fx.ResultTags(`name:"foo"`)
//    )
//  )
```